src/ripple/conditions/rsa_sha256.h

```diff
@@ -0,0 +1,270 @@
+
+        if (ret == -1)
+            return false;
+
+        BN_bn2bin (key->n, modulus_.alloc(BN_num_bytes (key->n)));
+        return true;
+    }
+
+    bool
+    validate_helper (RSA* key, Slice const& message) const
+    {
+        int const keySize = RSA_size(key);
+        if (!checkModulus (keySize))
+            return false;
+
+        Buffer buf;
+
+        auto ret = RSA_public_decrypt(
+            keySize,
+            signature_.data(),
+            buf.alloc (keySize),
+            key,
+            RSA_NO_PADDING);
+
+        if (ret == -1)
+            return false;
+
+        sha256_hasher h;
+        h (message.data(), message.size());
+        auto digest = static_cast<sha256_hasher::result_type>(h);
+
+        return RSA_verify_PKCS1_PSS(key, digest.data(), EVP_sha256(), buf.data(), -1) == 1;
+    }
+
+public:
+    rsa_sha256_t () = default;
+
+    condition_t
+    condition() const override
+    {
+        std::vector<std::uint8_t> m;
+        m.reserve (1024);
+
+        oer::encode_octetstring (
+            modulus_.size(),
+            modulus_.begin(),
+            modulus_.end(),
+            std::back_inserter(m));
+
+        sha256_hasher h;
+        h (m.data(), m.size());
+
+        condition_t cc;
+        cc.type = type();
+        cc.requires = requires();
+        cc.fulfillment_length =
+            oer::predict_octetstring_size (modulus_.size()) +
+            oer::predict_octetstring_size (signature_.size());
+        cc.fingerprint = static_cast<sha256_hasher::result_type>(h);
+
+        return cc;
+    }
+
+    std::uint16_t
+    type () const override
+    {
+        return condition_rsa_sha256;
+    }
+
+    std::uint32_t
+    requires () const override
+    {
+        return feature_rsa_pss | feature_sha256;
+    }
+
+    Buffer
+    payload() const override
+    {
+        Buffer b (
+            oer::predict_octetstring_size (modulus_.size()) +
+            oer::predict_octetstring_size (signature_.size()));
+
+        auto out = oer::encode_octetstring (
+            modulus_.size(),
+            modulus_.begin(),
+            modulus_.end(),
+            b.begin());
+
+        oer::encode_octetstring (
+            signature_.size(),
+            signature_.begin(),
+            signature_.end(),
+            out);
+
+        return b;
+    }
+
+    bool
+    validate (Slice const& data) const override
+    {
+        RsaKey rsa (RSA_new());
+
+        rsa->n = BN_new();
+        BN_bin2bn(modulus_.data(), modulus_.size(), rsa->n);
+
+        rsa->e = BN_new();
+        BN_set_word (rsa->e, 65537);
+
+        return validate_helper (rsa.get(), data);
+    }
+
+    /** Sign the given message with an RSA key */
+    bool sign (
+        std::string const& key,
+        Slice const& message)
+    {
+        // FIXME: Ugly const_cast/reinterpret_cast
+        auto bio = BIO_new_mem_buf(
+            const_cast<void*>(static_cast<void const*>(key.data())),
```

Comment:
Unfortunately, it is. The documentation for `BIO_new_mem_buf` suggests that the function accepts a `void const*`, but it requires a `void*` on two of my machines.